File: ScanProject/PyAnalysis/SonarPy/Medium/636-permutations-ii.py
#https://leetcode.com/problems/permutations-ii/
#Given a collection of numbers, nums, that might contain duplicates, return all possible unique permutations in any order.


class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """ 
        def dfs(nums, path, res):
            if not nums:
                res.append(path)
                return
            for i in range(len(nums)):
                if i > 0 and nums[i] == nums[i-1]:
                    continue
                dfs(nums[:i] + nums[i+1:], path + [nums[i]], res)

        res = []
        nums.sort()
        self.dfs(nums, [], res)
        return res
        

    

File: ScanProject/JSAnalysis/SonarJS/Medium/636-permutations-ii.js
/**https://leetcode.com/problems/permutations-ii/ */
//Given a collection of numbers, nums, that might contain duplicates, return all possible unique permutations in any order.


/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    let result = [];
    if(nums.length == 0){
        return result;
    }
    nums.sort((a, b) => a - b);
    let current = [];
    let visited = new Array(nums.length);
    helper(nums, current, result, visited);
    return result;
    
    
};
function helper(nums, current, result, visited){
    if(current.length == nums.length){
        result.push([...current]);
        return;
    }
    for(let i = 0; i < nums.length; i++){
        if(visited[i]){
            continue;
        }
        if(i > 0 && nums[i] == nums[i - 1] && !visited[i - 1]){
            continue;
        }
        current.push(nums[i]);
        visited[i] = true;
        helper(nums, current, result, visited);
        current.pop();
        visited[i] = false;
    }
}

    